### ISSUE TITLE
Submissions via e-mail with more than X MB can't be sent

### ISSUE THREAD

**Issue by alextreme** [open-formulieren/open-forms#3326]
### Product versie / Product version

latest / 2.3.0-alpha.0

### Omschrijf het probleem / Describe the bug

As noted with two gemeentes (Venray 68, HadM 56), forms with a multipleupload-widget or multiple single-file widgets can't easily be limited to a max total submission size.

As email submissions larger than 20MB are often undeliverable (yubin log status = 'failure') due to their size it either requires very low per-file size limits or daily manual checking in the OF admin for submissions with a total amount of files >20MB. It's also not possible to limit a form's submission size to a total size via the E-mailregistration plugin.

Of course the help-text of 'attach_files_to_email' does mention that care has to be taken to ensure the total file upload size isn't exceeded. However this warning is hidden behind a [?] in the configuration singleton and will be overlooked in most cases. It's also something which feels unfair (especially for large forms with many steps) and is unnecessary to ask of an average form designer. It also limits the usefulness of Open Forms as it means that files need to be sent out-of-bounds (via WeTransfer for instance) even though Open Forms is perfectly capable.

I propose that a max MB email size configuration option (default=20) is added to the 'E-mailregistratieconfiguratie'. Only e-mail submissions smaller than the configured size are then sent with attachments. Submissions with a total size of attachments larger than the configured size are then automatically not sent with attachments (as if attach_files_to_email = False), and automatically get a textual explaination stating where in OF the attachments of the submission can be found.

This avoids unnecessary incidents where submissions aren't received via e-mail and are overlooked.

### Stappen om te reproduceren / Steps to reproduce

1. Create a form that allows multiple files per submission
2. Configure emailregistration for the form
3. Send in a form submission >20MB
4. Check via Yubin log if the submission could be sent or not (depends per SMTP server)

### Verwacht gedrag / Expected behavior

Each succesful submission should be sent via e-mail, either with or without attachments. 

### Screen resolution

None

### Device

None

### OS

None

**Comment by joeribekker** [comment]
Including the uploads as attachments in the e-mail is discouraged for exactly the reasons you mention, and the default configuration is also to NOT attach them. Instead, the e-mail contains download links to the files for which you must have an admin account (that's better than a description where you can find the files 😉 ).

So I'm tempted to mark this as a wontfix - the requested complexity seems excessive to honour the problems that arise from not following recommended practices.

(written by Sergei, but I'll double check with Joeri and possibly have a spicy conversation)

**Comment by joeribekker** [comment]
Refinement: Real Joeri here: Most important is that the emails are not blocked without notice (other than logs). Joeri takes this back to the stakeholders to see what behaviour they want.

**Comment by alextreme** [comment]
Discussed this with Laurens, due to the fact that attachments are base64-encoded we would recommend that a max MB email size configuration option is set to 15 MB by default (base64-binary encoding increases size by ~30%, putting the post-MIME file size at max 19.5MB). 

If this option isn't possible we would recommend to disable sending the attachments via e-mail. In our eyes this feature should work consistently (every succesful submission generates a received e-mail) or it should not be given as an option.

**Comment by sergei-maertens** [comment]
> If this option isn't possible we would recommend to disable sending the attachments via e-mail. In our eyes this feature should work consistently (every succesful submission generates a received e-mail) or it should not be given as an option.

This is what we had and then we had stakeholders asking for exceptions because they do automatic processing of e-mails.

The max e-mail size setting also depends on which SMTP provider you are using. Some might bork at 20MB, others may bork at 10MB and it all depends on how you're sending those e-mails (e.g. if you're going through a mail server of the customer itself different rules may apply).

These size rules are arbitrary anyway - while the attachment sizes can be limited and you can account for the base64 inflation, you have no idea what the size of the HTML + plain text will be, since those are templates that are left to the form designer. This can still occasionally trip you over the limit (say that the combined attachments in base64 add up to 19.99MB and then you have 0.02 MB of text/HTML -> the SMTP server will reject this because it exceeds 20MB).

If we do anything about this, then we do it right and remove any guesswork from the equation (what's under our control, anyway), since any exceptional failures that only occur 0.001% of the time are even harder to debug and deal with.

**Comment by alextreme** [comment]
Mail and attachments are indeed annoying. However I don't see why removing 99.999% of the problem (of admins having to check all submissions manually on a daily basis) wouldn't be seen as an improvement. Perfect is the enemy of good imho.

But was we strive for perfect: For the 0.001% you could always solve this with a daily check to send a notification to the admin users with an overview of emails that couldn't be delivered via Yubin, we have a similar mechanism for other projects (i've hooked up one to Uptimerobot via a health check). The 0.001% indeed occurs but as checking the Yubin messages which haven't been succesfully delivered and communicating this daily is sufficient in my experience.

We can keep this parked for when Joeri is back and I don't think we've heard back from the affected municipalities yet. So your call if/what you want to do with this.

**Comment by sergei-maertens** [comment]
> However I don't see why removing 99.999% of the problem (of admins having to check all submissions manually on a daily basis) wouldn't be seen as an improvement. Perfect is the enemy of good imho.

Because in this particular case if you miss one or two e-mails and citizens requests aren't being processed, that's a very shitty experience for the actual "customer", while with the current scale of the problem the shitty experience is shifted to the people responsible for processing the requests. I do understand your point, but thinking about the technical amount of work needed to serve 99,9% I feel like we'd better invest all that energy in getting 100%.

> But was we strive for perfect: For the 0.001% you could always solve this with a daily check to send a notification to the admin users with an overview of emails that couldn't be delivered via Yubin, we have a similar mechanism for other projects (i've hooked up one to Uptimerobot via a health check).

Yeah that's something I want anyway for better traceability and it would be achieved by adding a custom e-mail header with the submission ID, so you can include it in all the (audit) logs of the submission and generate reports for that. But imo that's a quality of life improvement that needs to happen anyway and is only tangentially related to this particular problem.

> We can keep this parked for when Joeri is back and I don't think we've heard back from the affected municipalities yet. So your call if/what you want to do with this.

I definitely want to hear from the stakeholders first - I hope the outcome is that we can just drop e-mail attachments alltogether and then this whole discussion will have been pointless anyway. Let's first figure out what people want before we start coming up with solutions - otherwise we end up with another solution that needs a problem ;)

**Comment by sergei-maertens** [comment]
I'm speccing out that ticket and putting it on the planning for 2.4.0
